File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;
class Solution {
    int[] dx = {0, 0, -1, 1};
    int[] dy = {1, -1, 0, 0}; // 상하좌우
    public int solution(int[][] maps) {
        int answer = 0;
        int[][] visited = new int[maps.length][maps[0].length];
        
        bfs(maps, visited);
        answer = visited[maps.length-1][maps[0].length-1];
        
        if(answer == 0){
            answer = -1;
        }
        
        return answer;
    }
    public void bfs(int[][] maps, int[][] visited){

        visited[0][0] = 1;
        Queue<int[]> queue = new LinkedList<>();
        queue.add(new int[]{0, 0});
        
        while(!queue.isEmpty()){
            int[] cur = queue.remove();
            int cx = cur[0];
            int cy = cur[1];
            
            for(int i = 0; i < 4; i++){
                int nx = cx + dx[i];
                int ny = cy + dy[i];
                
                if(nx<0 || nx>maps.length-1 || ny<0 || ny>maps[0].length-1){
                    continue;
                }
                
                 if(visited[nx][ny] == 0 && maps[nx][ny] == 1){
                    visited[nx][ny] = visited[cx][cy] + 1;
                    queue.add(new int[]{nx, ny});
                }
            }
        }
        
    }
    
    
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    answer = 0
    visited = [[False]*len(maps[0]) for _ in range(len(maps))]
    q = deque()
    q.append((0,0))
    dr = [0,0,-1,1]
    dc = [1,-1,0,0]
    visited[0][0] = True
    
    while q:
        r,c = q.popleft()
        for i in range(4):
            nr = dr[i]+r
            nc = dc[i]+c
            if 0 <= nr <len(maps) and 0<= nc <len(maps[0]) and not visited[nr][nc] and maps[nr][nc]==1:
                visited[nr][nc] = True
                q.append((nr,nc))
                maps[nr][nc] = maps[r][c] + 1
    
    if maps[len(maps)-1][len(maps[0])-1] == 1:
        return -1
    
    else:
        return maps[len(maps)-1][len(maps[0])-1]
               
               